partiql-parser/src/result.rs

```diff
@@ -4,33 +4,140 @@
 use std::fmt;
+use std::fmt::Formatter;
 use thiserror::Error;
 
-/// Position in the source for an error.
+/// A line and column location.
+///
+/// This value is one-based, as that is how most people think of lines and columns.
+///
+/// ## Example
+/// ```
+/// # use partiql_parser::prelude::*;
+/// println!("Beginning of a document: {}", LineAndColumn::at(1, 1));
+/// ```
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct LineAndColumn(pub usize, pub usize);
+
+impl LineAndColumn {
+    /// Constructs a [`LineAndColumn`].
+    #[inline]
+    pub fn at(line: usize, column: usize) -> Self {
+        Self(line, column)
+    }
+
+    /// Returns a [`LineAndColumn`] that moves to the given position relative
+    /// to this one as a sort of "origin."
+    ///
+    /// Note that this positioning is 1-based, so moving `(1, 1)` to `(1, 1)` is a no-op.
+    ///
+    /// ## Examples
+    /// ```
+    /// # use partiql_parser::prelude::*;
+    /// assert_eq!(
+    ///     LineAndColumn::at(1, 1),
+    ///     LineAndColumn::at(1, 1).move_relative(LineAndColumn::at(1, 1))
+    /// );
+    /// ```
+    ///
+    /// ```
+    /// # use partiql_parser::prelude::*;
+    /// assert_eq!(
+    ///     LineAndColumn::at(1, 2),
+    ///     LineAndColumn::at(1, 1).move_relative(LineAndColumn::at(1, 2))
+    /// );
+    /// ```
+    ///
+    /// ```
+    /// # use partiql_parser::prelude::*;
+    /// assert_eq!(
+    ///     LineAndColumn::at(5, 10),
+    ///     LineAndColumn::at(5, 7).move_relative(LineAndColumn::at(1, 4))
+    /// );
+    /// ```
+    ///
+    /// ```
+    /// # use partiql_parser::prelude::*;
+    /// assert_eq!(
+    ///     LineAndColumn::at(21, 2),
+    ///     LineAndColumn::at(2, 15).move_relative(LineAndColumn::at(20, 2))
+    /// );
+    /// ```
+    pub fn move_relative(self, location: LineAndColumn) -> Self {
+        match (self, location) {
+            (LineAndColumn(base_line, base_column), LineAndColumn(dest_line, dest_column)) => {
+                let diff_line = dest_line - 1;
+                if diff_line > 0 {
+                    // we're moving lines, adjust the line and take the target column as-is
+                    LineAndColumn::at(base_line + diff_line, dest_column)
+                } else {
+                    // same line from base, adjust only the column
+                    let diff_column = dest_column - 1;
+                    LineAndColumn::at(base_line, base_column + diff_column)
+                }
+            }
+        }
+    }
+}
+
+impl From<(usize, usize)> for LineAndColumn {
+    fn from(line_and_column: (usize, usize)) -> Self {
+        let (line, column) = line_and_column;
+        Self::at(line, column)
+    }
+}
+
+impl fmt::Display for LineAndColumn {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "line {}, column {}", self.0, self.1)
+    }
+}
+
+/// A possible position in the source.
 #[derive(Copy, Clone, Debug, Eq, PartialEq)]
 pub enum Position {
```

Comment:
Do you want EOF?

Reply:
Do you have a use case where distinguishing `EndOfFile` as a position is needed rather than the line/column that happens to be at the `EndOfFile`?  I do think we might need a `UnexpectedEndOfFile` type error case, but not sure about this one, an example might help.

Reply:
doing EOF in the error + the L+C makes sense!